### ISSUE TITLE
FlxCamera follow jitters

### ISSUE THREAD

**Issue by FlixelCommunityBot** [FlixelCommunity/flixel#5]
Issue [#227](https://github.com/AdamAtomic/flixel/issues/227) by: [volvis](http://github.com/volvis)

I noticed an ever slight jitter around my playable character when the camera was tracking it. Imagine the sprite jumping a pixel in advance and back the second frame and you get a good idea of the effect I'm talking about.

Poking around the update call in FlxCamera, rounding the coordinates of the game character seemed to fix the issue.

``` as3
var targetX:Number = int(target.x) + ((target.x > 0)?0.001:-0.001);
var targetY:Number = int(target.y) + ((target.y > 0)?0.001: -0.001);
```

Can't really tell if it's a bug or not, but I figured I'd toss it out in the open.

**Comment by IQAndreas** [comment]
This bug is also mentioned in the following thread:
[Fix for jittery camera movement (Flixel 2.5)](http://forums.flixel.org/index.php/topic,4543.0.html)

It seems like Krix has been hard at work making additional changes than just the one fix listed in this issue's comments.
https://github.com/krix/CameraScrollFix

**Comment by Dovyski** [comment]
I've noticed Krix work. I analyzed his commit log and there seems to be a huge change in `FlxSprite` regarding `framePixels` (https://github.com/krix/CameraScrollFix/commit/5af35473b652fcc69acecc7d31ee2e86f536513e). The property is changed from `BitmapData` to `Bitmap`. It will break lots of codes out there, I imagine.

It would be great if we could add the camera fix without touching too many other classes.

**Comment by Gama11** [comment]
Apparently, replacing those two lines with:

``` actionscript
var targetX:Number = FlxU.ceil(target.x + ((target.x > 0)?0.0000001:-0.0000001));
var targetY:Number = FlxU.ceil(target.y + ((target.y > 0)?0.0000001:-0.0000001));
```

like mentioned in the thread seems to fix all the issues except for the `STYLE_LOCKON` and `SHAKE_BOTH_AXES` camera modes - I just tested that, as funnily enough, I ran into the exact same problem a few hours ago. So, I guess before not fixing anything at all, this fix should be made to make everything but two modes work smoothly.

**Comment by IQAndreas** [comment]
>  The property is changed from BitmapData to Bitmap. It will break lots of codes out there, I imagine.

I noticed the same, and I have been meaning to send an email to him.

The thing is, in every instance where he accesses the `framePixels` object, he uses `framePixels.bitmapData`. I really don't understand why he made the change, as it is not at all necessary, and doesn't have anything to do with the actual "camera scroll fix". Unless Krix had some other reason for doing so, those changes should be filtered out and not merged in to this project.

Krix also has several changes made to his personal Flixel branch, but since the changes in the two projects use different commit identifiers, I have yet to filter which ones have been applied in both repositories. 

I was hoping @Krix would be willing to do it if he still remembers the changes he made (you know how that is, picking up an old project after a few months, it's often as if someone else wrote the code).

**Comment by krix** [comment]
Hi there!

As far as I remember, I made the change due to a memory leak issue with the draw() function in flash > 10.
There were no side effects in any of my projects, but I understand if you don't want to adobt this fix.

However, if you just want to fix the jittering, you should do the following changes:
https://github.com/krix/flixel/commit/67897031bd78d4f55fda8057bc4fc49e915c20eb
https://github.com/krix/flixel/commit/ba58e6d52f5631990df0c7283ef291b54e1af16d
and maybe this too: https://github.com/krix/flixel/commit/7fbc822359dfdfd6d74d9b9eaf86e0699c1cb0da

I hope this helps.

Cheers,
Dirk

**Comment by IQAndreas** [comment]
> As far as I remember, I made the change due to a memory leak issue with the draw() function in flash > 10.

Ah, magnificent! :) Thanks for the quick clarification, Dirk.

In case anyone was looking for more information on the bug:
[Drawing API Memory-Leak in Flash 10.1](http://blog.pixelpracht.net/?p=122)

The good news is, it seems [this issue was resolved July 2011](https://bugs.adobe.com/jira/browse/FP-4812). It doesn't say which version of Flash Player it was fixed in, but I think it's safe to say, that version is in "standard use" right now. (I was going to post a link to the Flash Player Version statistics here, but alas, I can't any updated stats).

**Comment by IQAndreas** [comment]
@krix Is this commit part of this bug, or is it referring to a different "jittering"? https://github.com/krix/flixel/commit/c5ca1cac1e2e531725abe7c9de46fb13be3c45a7

_EDIT:_ Never mind, I see the change was undone. https://github.com/krix/flixel/commit/b34c606e016a99ca98bf96717a5bf1dda81fbf19

**Comment by IQAndreas** [comment]
Fixed in 1f19b26596e3a6762982ad1ea20edc20fbea227e
